File: packages/hydris/src/error/HttpException.ts
export class HttpException extends Error {
    constructor(
        public readonly status: number,
        message?: string,
    ) {
        super(message ?? `HTTP ${status}`);
        this.name = new.target.name;
    }
}

export class BadRequestException extends HttpException {
    constructor(message?: string) {
        super(400, message);
    }
}

export class UnauthorizedException extends HttpException {
    constructor(message?: string) {
        super(401, message);
    }
}

export class ForbiddenException extends HttpException {
    constructor(message?: string) {
        super(403, message);
    }
}

export class NotFoundException extends HttpException {
    constructor(message?: string) {
        super(404, message);
    }
}

export class MethodNotAllowedException extends HttpException {
    constructor(
        public readonly allowed: string[] = [],
        message?: string,
    ) {
        super(405, message);
    }
}

export class ValidationException extends HttpException {
    constructor(message?: string) {
        super(422, message);
    }
}

File: packages/hydris/Hydris/Error/HttpException.cs
namespace Hydris.Error;

public class HttpException : Exception {
    public int Status { get; }

    public HttpException(int status, string? message = null)
        : base(message ?? ReasonPhrase(status)) {
        Status = status;
    }

    private static string ReasonPhrase(int status) => status switch {
        400 => "Bad Request",
        401 => "Unauthorized",
        403 => "Forbidden",
        404 => "Not Found",
        405 => "Method Not Allowed",
        422 => "Unprocessable Entity",
        500 => "Internal Server Error",
        _ => $"HTTP {status}",
    };
}

public sealed class BadRequestException(string? message = null) : HttpException(400, message);

public sealed class UnauthorizedException(string? message = null) : HttpException(401, message);

public sealed class ForbiddenException(string? message = null) : HttpException(403, message);

public sealed class NotFoundException(string? message = null) : HttpException(404, message);

public sealed class ValidationException(string? message = null) : HttpException(422, message);

public sealed class MethodNotAllowedException(IReadOnlyList<string> allowed, string? message = null)
    : HttpException(405, message) {
    public IReadOnlyList<string> Allowed { get; } = allowed;
}
